### ISSUE TITLE
Kernels in shared libraries fail to execute

### ISSUE THREAD

**Issue by fkruggel** [ROCm-Developer-Tools/HIP#1055]
Hi there,

I am unable to run kernels in shared libraries. Please, refer to the attached boiled-down example. The file test.cpp contains a driver that calls the function testKernel() in file gpu.cpp.
1. When I link both files directly, the kernel executes:
     hipcc test.o gpu.o -l test; ./test
2. Now place gpu.o into a shared library:
    hipcc -shared -Wl,-soname,libtest.so -o libtest.so gpu.o; hipcc test.o -L. -ltest -o test; ./test
    results in a segfault at hsa_init(). I found that this library has several global variables that are
    not initialized by the time hsa_init() is called.
3. Use g++ instead of hipcc:
    g++ -shared -Wl,-soname,libtest.so -o libtest.so gpu.o; hipcc test.o -L. -ltest -o test; ./test
    Here, hsa_init() succeeds, but the kernel launch fails.
All hip examples run fine (but none uses shared libraries). My HIP environment is freshly compiled as of this week. May I kindly ask you for advice re: building shared libraries?

Thanks, Frithjof

[test.txt](https://github.com/ROCm-Developer-Tools/HIP/files/3118370/test.txt)
[gpu.txt](https://github.com/ROCm-Developer-Tools/HIP/files/3118371/gpu.txt)
[hipInfo.txt](https://github.com/ROCm-Developer-Tools/HIP/files/3118376/hipInfo.txt)

**Comment by fkruggel** [comment]
With a fresh install of ROC version 2.4, I can confirm that tests 1. and 2. work. 
Running an executable built with a g++-constructed shared library (test 3) does not work, because the function function_names() in line 107 of functional_grid_launch.hpp does not find the kernel function signature, although it is present in the executable.

**Comment by SarbojitAMD** [comment]
**Summary** : Initial investigation looks like **not a bug**, will wait for submitter's confirmation.

I am able to launch kernel from shard library without any crash, I used the same set of files/code (gpu.txt & test.txt) that submitter had attached here. I used below commands to compile the code

**Steps:**
1. hipcc -fPIC $(INC) -c gpu.cpp     // will generate gpu.o object file
2. hipcc -shared -o libgpu.so gpu.o $(LIB) -i:libhip_hcc.so   // will generate shared library name libgpu.so
3. g++ $(INC) -D __HIP_PLATFORM_HCC__ -c test.cpp -std=c++11 // Macro is required to force HIP to take HCC path for required linkings, c++ standard you can choose anything I have chosen as HIP code follows c++11
4. hipcc -o test test.o libgpu.so

Note : For temporary you can use LD_LIBRARY_PATH to point to libgpu.so location so that "./test" will be able to invoke required APIs. Permanent solution is to copy the library under /usr/lib/

**Difference in ABOVE steps and CURRENT steps:**

1. I could not see libhip_hcc.so library was used any where in submitter's steps which should have given failure during linking. 
2. Linking was done during host code build, which I feel not needed as object code are getting built hence linking in last step make more sense. 

Please let me know in case you need more help.

**Comment by SarbojitAMD** [comment]
I have already mentioned steps that worked for me with same code files shared by submitter. hence closing it.

**Comment by fkruggel** [comment]
Thanks for your update. Please note that I already confirmed
that my test cases 1 and 2 worked after an update to ROC
version 2.4 (on May 14).

However, I still find that generating a shared library with gcc
(my test case 3) does not work. Note that this case is not tested
or covered in your answer.

It might be helpful to learn which ROC release you are using.

Thanks and best wishes, Frithjof

**Comment by Sarbojit2019** [comment]
Thanks for your response, I really wanted to understand your intension in test 3. That's why I have given my working steps with some extra comments added. If you feel that is not what you want to achieve then please clearly call out your intension so that I can help you with steps.
 
From your mentioned steps looks like you wanted to build GPU code (kernel) using g++ which is not possible. You have to use hipcc or nvcc to build device code. Having said that we can always create/build shared library with device code using hipcc and build host code in g++, finally link then using hipcc as g++ does not understand device instruction.

I am using ROC 2.6 version

**Comment by fkruggel** [comment]
Thanks for getting back to me. I apologize if the intentions
for my examples were not fully clear.

> From your mentioned steps looks like you wanted to build GPU code
> (kernel) using g++ which is not possible.
It is perfectly clear to me that hipcc is needed to build device
code. hipcc was used for all compilation steps (.cpp -> o.). Note
that I did not mention compilation my examples.
If hipcc is needed for GPU code generation during a later step
(i.e., shared library build or linking), please, state so, because
then I would understand the issue.

> I really wanted to understand your intension in test 3.
The intention was to link a program against a shared library built
by hipcc (test 2) or g++ (test 3) from compiled objects. As said,
the program built in test 2 executes correctly, while the program
from test 3 fails to execute. From inspection, I find GPU code in
both executables. While executing test 2, the GPU code is found
(and executed), for test 3, the GPU code is not found.

So I assume that during the shared library build, object modules
may be flagged/marked in a different way such that the hipcc runtime
can (test 2) or cannot (test 3) find the GPU code.

Just for the sake of clarity, please, let me rephrase:
1. I can run GPU code contained in a shared library if the
GPU kernel is compiled by hipcc, the library built by hipcc,
and the program linked by hipcc.
2. I cannot run GPU code contained in a shared library if the
GPU kernel is compiled by hipcc, the library built by g++,
and the program linked by hipcc or g++.

Please note that all three test cases run fine on NVIDIA cards
(using nvcc & g++).

Thanks again and best wishes, Frithjof

**Comment by SarbojitAMD** [comment]
Thanks you for explaining the intent 

You guessed it right, hipcc is needed for GPU code generation during a later stages e.g. shared library building or linking. It is because hipcc during compile only emits LLMV IR and later during linking it generates ISA hence creating shared library using hipcc is must on AMD platform.

**Comment by fkruggel** [comment]
I suggest to add this information to the documentation.
Thanks for your clarification and help. Case closed.
